app/models/user.rb

```diff
@@ -25,80 +29,165 @@ def to_s
   def set_groups_to_impersonate(grps)
-    @groups_to_impersonate = grps
+    if grps.blank? # blank is falsy values like: nil, '', []
+      # remove impersonation (see #groups below)
+      @role_cache = {}
+      @groups_to_impersonate = nil
+    elsif grps.is_a? Array
+      @role_cache = {}
+      @groups_to_impersonate = grps
+    elsif grps.is_a? String
+      @role_cache = {}
+      @groups_to_impersonate = [grps]
+    else
+      raise ArgumentError.new("Cannot accept #{grps.class} argument")
+    end
   end
 
   def groups
-    return @groups_to_impersonate if @groups_to_impersonate
-    perm_keys = ["sunetid:#{login}"]
-    return perm_keys unless webauth && webauth.privgroup.present?
-    perm_keys + webauth.privgroup.split(/\|/).collect { |g| "workgroup:#{g}" }
+    return @groups_to_impersonate unless @groups_to_impersonate.blank?
+    @groups ||= begin
+      perm_keys = ["sunetid:#{login}"]
+      return perm_keys unless webauth && webauth.privgroup.present?
+      perm_keys + webauth.privgroup.split(/\|/).map {|g| "workgroup:#{g}"}
+    end
   end
 
   # @return [Boolean] is the user a repository wide administrator
   def is_admin
-    !(groups & ADMIN_GROUPS).empty?
-  end
-
-  # @return [Boolean] is the user a repository wide viewer
-  def is_viewer
-    !(groups & VIEWER_GROUPS).empty?
+    !(groups & ADMIN_GROUPS).blank?
   end
 
   # @return [Boolean] is the user a repo wide manager
   def is_manager
-    !(groups & MANAGER_GROUPS).empty?
+    !(groups & MANAGER_GROUPS).blank?
+  end
+
+  # @return [Boolean] is the user a repository wide viewer
+  def is_viewer
+    !(groups & VIEWER_GROUPS).blank?
   end
 
   # The convention is for boolean is_XYZ methods to be interrogative (end in "?")
   alias_method :is_admin?,   :is_admin
-  alias_method :is_viewer?,  :is_viewer
   alias_method :is_manager?, :is_manager
+  alias_method :is_viewer?,  :is_viewer
 
   def login
     webauth ? webauth.login : sunetid
   end
 
-  ##
+  # Validate access permissions using Argo user roles. If that fails, check the
+  # APO role permissions.  If the object is an APO, first check permissions on
+  # the governing APO of this APO object. As a last resort, check the object
+  # roles when this object is an APO.
+  # @param dor_object [String|Dor::Base] Accepts a DRUID string or
+  #                                      any subclass of Dor::Base
+  # @return [Boolean]
+  def can_manage_object?(dor_object)
+    dor_object = parse_object(dor_object)
+    pid = dor_object.pid
+    apo_pid = dor_object.admin_policy_object.pid
+    manager_access = is_admin || is_manager
+    manager_access ||= dor_object.can_manage_item?(roles(apo_pid))
+    manager_access ||= dor_object.can_manage_content?(roles(apo_pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_item?(roles(pid))
+    manager_access ||= dor_object.is_a?(Dor::AdminPolicyObject) && dor_object.can_manage_content?(roles(pid))
```

Comment:
i think the above two lines might be a bug.

my understanding of the above is as follows:
* if the user is a repo wide admin or manager (`is_admin` or `is_manager` is true), return true.
* if the user can manage items or manage content for things governed by `apo_pid`, return true.
* this is the part that i believe is a bug:  if the user can manage items or content for things governed by `pid`, in the case that `pid` represents an APO, return true.  which is to say, if you can manage things governed by an APO, you can manage the APO itself.

if this is the case, i can see how it is analogous to what was asked for in the issue for which this PR was created:  the desired behavior was that if you can see an item, you should be able to see the APO which governs it.  more accurately relative to this implementation, if an APO might allow you to see an object, you should be able to see that APO.

however, that's quite different from saying that if an APO allows you to manage an object, you should be able to manage that APO.  i don't think that was implied in the issue description, and i don't think it's desirable behavior, but that's how this code currently reads to me.

in looking at the tests, these lines appear not to be exercised.  nothing seems to specifically test the case where `dor_object.is_a?(Dor::AdminPolicyObject)` evaluates to true.